src/EditorFeatures/Core/Implementation/AddImports/AddImportsPasteCommandHandler.cs

```diff
@@ -0,0 +1,120 @@
+
+namespace Microsoft.CodeAnalysis.Editor.Implementation.AddImports
+{
+    [Export]
+    [Export(typeof(ICommandHandler))]
+    [ContentType(ContentTypeNames.RoslynContentType)]
+    [Name(PredefinedCommandHandlerNames.AddImportsPaste)]
+    // Order is important here, this command needs to execute before PasteTracking
+    // since it may modify the pasted span. Paste tracking dismisses if 
+    // the span is modified. It doesn't need to be before FormatDocument, but
+    // this helps the order of execution be more constant in case there 
+    // are problems that arise. This command will always execute the next
+    // command before doing operations.
+    [Order(After = PredefinedCommandHandlerNames.PasteTrackingPaste)]
+    [Order(Before = PredefinedCommandHandlerNames.FormatDocument)]
+    internal class AddImportsPasteCommandHandler : IChainedCommandHandler<PasteCommandArgs>
+    {
+        public string DisplayName => EditorFeaturesResources.Add_Missing_Imports_On_Paste;
+
+        private readonly IThreadingContext _threadingContext;
+
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public AddImportsPasteCommandHandler(IThreadingContext threadingContext)
+            => _threadingContext = threadingContext;
+
+        public void ExecuteCommand(PasteCommandArgs args, Action nextCommandHandler, CommandExecutionContext executionContext)
+        {
+            // Capture the pre-paste caret position
+            var caretPosition = args.TextView.GetCaretPoint(args.SubjectBuffer);
+            if (!caretPosition.HasValue)
+            {
+                return;
+            }
+
+            nextCommandHandler();
+
+            if (!args.SubjectBuffer.CanApplyChangeDocumentToWorkspace())
+            {
+                return;
+            }
+
+            // Don't perform work if we're inside the interactive window
+            if (args.TextView.IsBufferInInteractiveWindow(args.SubjectBuffer))
+            {
+                return;
+            }
+
+            if (!args.SubjectBuffer.GetFeatureOnOffOption(FeatureOnOffOptions.AddImportsOnPaste))
+            {
+                return;
+            }
+
+            // Create a tracking span from the pre-paste caret position that will grow as text is inserted.
+            var trackingSpan = caretPosition.Value.Snapshot.CreateTrackingSpan(caretPosition.Value.Position, 0, SpanTrackingMode.EdgeInclusive);
+
+            // Applying the post-paste snapshot to the tracking span gives us the span of pasted text.
+            var snapshotSpan = trackingSpan.GetSpan(args.SubjectBuffer.CurrentSnapshot);
+            var textSpan = TextSpan.FromBounds(snapshotSpan.Start, snapshotSpan.End);
+
+            AddMissingImportsForPaste(args, executionContext, textSpan);
+        }
+
+        public CommandState GetCommandState(PasteCommandArgs args, Func<CommandState> nextCommandHandler)
+            => nextCommandHandler();
```

Comment:
move this above the method before it.

Reply:
Why?

Reply:
Two reasons:

First, command handlers generally fit as a pair of "is this command available?" and "execute the command if so".  So keeping the methods ordered that way fits nicely.  

Second, you currently have the methods like so:

```
void DoSomething1() { DoSomething2(); }
void Unrelated() { }
void DoSomething2() { }
```

This means that the two related methods are unnecessarily separated.  Reordering gives a much easier to scan and understand:


```
void Unrelated() { }
void DoSomething1() { DoSomething2(); }
void DoSomething2() { }
```

Reply:
The first reason I guess works? It's not something that's enforced in the various `ICommandHandler` implementations we have. 

The second reason will happen a lot with the requirement of ordering based on accessibility of the method though right? It seems more cumbersome to specify that only in cases where a certain number of public methods are available should you order them closest to the private/protected methods they call.

Reply:
> The second reason will happen a lot with the requirement of ordering based on accessibility of the method though right

We don't have a requirement like that :-)

Reply:
Well technically we don't have any guidelines on method ordering in [the coding style](https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/coding-style.md) documentation. If I leave it in the way Visual Studio added for "Implement interface" then it's correct according to `The general rule we follow is "use Visual Studio defaults".` 

I still stand by my point that you're enforcing a rule that doesn't exist and isn't enforced across the code base. We can argue in some other venue though, since I doubt this is conducive to the PR and I've already moved it.